src/qt/importkeysdialog.cpp

```diff
@@ -0,0 +1,164 @@
+#include "validation.h"
+#include "wallet/wallet.h"
+#include "walletmodel.h"
+#include "util.h"
+
+#include <QThread>
+#include <QDebug>
+
+/* Object for executing key import commands in a separate thread.
+*/
+class ImportKeyExecutor : public QObject
+{
+    Q_OBJECT
+
+public Q_SLOTS:
+    void rescan(CWallet*, CBlockIndex*);
+
+Q_SIGNALS:
+    void rescanWallet(CWallet*, CBlockIndex*);
+};
+
+#include "importkeysdialog.moc"
+
+void ImportKeyExecutor::rescan(CWallet* pwallet, CBlockIndex* genesisBlock)
+{
+    qWarning() << "started import key thread";
+    pwallet->UpdateTimeFirstKey(1);
+    pwallet->ScanForWalletTransactions(genesisBlock, true);
+    qWarning() << "quitting import key thread";
+    QObject::thread()->quit();
+}
+
+ImportKeysDialog::ImportKeysDialog(const PlatformStyle *_platformStyle, QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::ImportKeysDialog),
+    platformStyle(_platformStyle)
+{
+    ui->setupUi(this);
+
+    /* Main elements init */
+    if (fPruneMode) {
+        ui->rescanCheckBox->setEnabled(false);
+    }
+}
+
+ImportKeysDialog::~ImportKeysDialog()
+{
+    thread.wait();
+    delete ui;
+}
+
+void ImportKeysDialog::on_okButton_clicked()
+{
+    if (importKey()) {
+        resetDialogValues();
+        accept();
+    };
+}
+
+void ImportKeysDialog::on_cancelButton_clicked()
+{
+    reject();
+}
+
+void ImportKeysDialog::on_resetButton_clicked()
+{
+    resetDialogValues();
+}
+
+bool ImportKeysDialog::importKey()
+{
+    const QString privateKey      = ui->privateKey->text();
+    const QString privateKeyLabel = ui->privateKeyLabel->text();
+    const bool rescan             = ui->rescanCheckBox->isChecked();
+
+    resetDialogValues();
+
+    CBitcoinSecret vchSecret;
+    bool fGood = vchSecret.SetString(privateKey.toStdString());
+    if (!fGood) {
+        vchSecret.SetString("");
+        ui->privateKeyImportTextMessage->setText(tr("Invalid private key; please check and try again!"));
+        return false;
+    }
+
+    CKey key = vchSecret.GetKey();
+    if (!key.IsValid()) {
+        vchSecret.SetString("");
+        ui->privateKeyImportTextMessage->setText(tr("Invalid private key; please check and try again!"));
+        return false;
+    }
+
+    CPubKey pubkey = key.GetPubKey();
+    assert(key.VerifyPubKey(pubkey));
+    CKeyID vchAddress = pubkey.GetID();
+
+    pwalletMain->MarkDirty();
+    pwalletMain->SetAddressBook(vchAddress, privateKeyLabel.toStdString(), "receive");
+
+    if (pwalletMain->HaveKey(vchAddress)) {
+        vchSecret.SetString("");
+        ui->privateKeyImportTextMessage->setText(
+            tr("Invalid address generated from private key; please check and try again!"
+        ));
+        return false;
+    }
+
+    pwalletMain->mapKeyMetadata[vchAddress].nCreateTime = 1;
+
+    if (!pwalletMain->AddKeyPubKey(key, pubkey)) {
+        vchSecret.SetString("");
+        ui->privateKeyImportTextMessage->setText(tr("Couldn't add private key; does it already exist?"));
```

Comment:
Are we sure that when a key already exists, we want to react differently than when it doesn't? IIRC this is why RPC doesn't throw an error on duplicate key import, because that could be used to expose which keys are in the wallet?

Reply:
Yeah, I'm not sure what the right balance is here. It'd be nice to give some suggestion of what the user could do to correct the problem, but I tried to phrase things vaguely enough that it's not leaking information.

Maybe something more like

  Couldn't add private key. This can be for several reasons:
  
   * the key already exists
   * the key is malformed
   * the wallet is malformed
 
I don't care much for that; it's a lot of text, it's vague, and it still suffers from the flaw you mention. Maybe "Failed to add private key" is the least worst option.

Reply:
Alright. I'm okay with "Failed to add private key" to unblock this for now. 

I will manually test every scenario I can think of between RPC and UI before merge, if I find any useful improvements we can address that in a subsequent PR.